substrate/primitives/trie/src/node_codec.rs

```diff
@@ -110,6 +110,10 @@ where
 			NodeHeader::Null => Ok(NodePlan::Empty),
 			NodeHeader::HashedValueBranch(nibble_count) | NodeHeader::Branch(_, nibble_count) => {
 				let padding = nibble_count % nibble_ops::NIBBLE_PER_BYTE != 0;
+				// data should be at least the size of the offset
+				if data.len() < input.offset {
+					return Err(Error::BadFormat)
+				}
```

Comment:
Alternative would be to do the check with let `offset = input.take(1)?.start`, then use offset instead of input.offset. (not exposing input.offset).

Reply:
Hey @cheme! Thanks for the feedback. I like your alternative solution but the reason I went for a simple `if` clause was because `take` increments the `offset` of `input`, so a big part of the code would have to be refactored to adapt to this which might not be worth it for a simple fix like this.

Nevertheless, I made a mistake in this PR since the correct check would be `data.len() < input.offset + 1` to make sure `data[input.offset]` does not panic 😅 I realized after testing and this PR was sadly already merged so I'm a bit ashamed, sorry about that. I already opened another PR (#6502) that fixes this

Reply:
yeah, would have needed a `peak` method (reader is implemented to be usable with scale api, iirc there another std trait that expose peak), but really not worth it.